### PR TITLE
Phoenix app template auth

### PR DESCRIPTION
**Pull request by Harleymckee**
Initial logged in logged out flow for RN template.

[WIP]
Things left todo:
1. clean up Typing a bit more
2. consider state mgmt alternatives to unstated (do we like Redux better? or just call it a day with unstated - it's worked out fine for me)
3. import base styles from https://github.com/revelrylabs/Harmonium-RN or by using Bryans design tokens in newest version of Harmonium
4. Switch Authentication to Oauth2 pattern 
- [x] use encrypted storage to store tokens

### CODE REVIEW

**Comment by Harleymckee** [comment]
- [ ] gitignore ./Pods

File: package.json [hunk 1]
```diff
@@ -25,5 +25,6 @@
   "bugs": {
     "url": "https://github.com/emin93/react-native-template-typescript/issues"
   },
-  "homepage": "https://github.com/emin93/react-native-template-typescript#readme"
+  "homepage": "https://github.com/emin93/react-native-template-typescript#readme",
```
Comment (Harleymckee) [inline package.json]: zap

File: template/src/components/Auth/Login.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      navigation: {navigate},
+    } = this.props
+
+    const user = {
+      password: this.state.password,
+      email: this.state.email,
+    }
+
+    if (
+      requiredFields(
```
Comment (Harleymckee) [inline template/src/components/Auth/Login.tsx]: hate this requiredFields helper

File: template/src/styles/components/containers.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import * as Colors from '../colors'
+import * as Spacing from '../spacing'
+
+export const pageWrapper = {
+  backgroundColor: Colors.lightestGray,
+  flex: 1,
+  width: fullWidth,
+}
+export const base = {
```
Comment (Harleymckee) [inline template/src/styles/components/containers.js]: Move Base styles to own repo.

File: template/src/containers/AuthContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  logInUser = async (user, navigate) => {
+    try {
+      await this.setState({isFetching: true})
+
+      const results = await post(`${global.apiUrl}/authenticate`, {data: user})
+
+      const attributes = results.data.attributes
+
+      await this.setState(
```
Comment (Dacello) [inline template/src/containers/AuthContainer.tsx]: huh didnt know you could `await` `setState`

File: template/src/navigation/SettingsStack.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import React from 'react'
+
+import {createStackNavigator} from 'react-navigation-stack'
+import BusyButton from '../components/BusyButton'
+import Feather from 'react-native-vector-icons/Feather'
+import {Text, ScrollView, View, SafeAreaView, AsyncStorage} from 'react-native'
```
Comment (Dacello) [inline template/src/navigation/SettingsStack.tsx]: nit: I prefer to have organize imports as external first, followed by aliased paths, followed by relative paths. You aren't doing any aliasing atm, and thats something that might be cool to add. We did it in TSS with the `metro-react-native-babel-preset` package (see https://github.com/revelrylabs/TSS/blob/develop/mobile/babel.config.js)

File: template/src/navigation/SettingsStack.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  {
+    initialRouteName: 'Settings',
+    defaultNavigationOptions: ({navigation}) => ({
+      title: 'Settings',
+      headerLeft: (
+        <Feather
+          style={{left: 10}}
+          name="menu"
+          size={32}
+          color="white"
```
Comment (Dacello) [inline template/src/navigation/SettingsStack.tsx]: `white` should prly come from `Colors`

File: template/src/styles/DrawerStyles.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    bottom: 2,
+    right: 10,
+  },
+  drawerHeaderEditText: {
+    color: 'white',
+    fontSize: Typography.bodyFontSizeLarge,
+    textAlign: 'center',
+  },
+  drawerContent: {
+    backgroundColor: 'white',
```
Comment (Dacello) [inline template/src/styles/DrawerStyles.ts]: grab from `Colors`

File: template/src/containers/AuthContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        {currentUser: {...attributes.user, token: attributes.access_token}},
+        () => {
+          AsyncStorage.multiSet([
+            ['token', this.state.currentUser.token],
```
Comment (Dacello) [inline template/src/containers/AuthContainer.tsx]: do we need to worry about https://tools.ietf.org/html/rfc8252 re: not storing sensitive information in async storage? App template seems like a good place to figure out how to do it correctly

File: template/App.tsx [hunk 9]
```diff
@@ -1,117 +1,69 @@
...
-    fontWeight: '600',
-    padding: 4,
-    paddingRight: 12,
-    textAlign: 'right',
-  },
-});
+const AppContainer = createAppContainer(AppNavigator)
+
+export default class App extends Component {
+  async componentDidMount() {
```
Comment (Dacello) [inline template/App.tsx]: reason this is async? But this concept of `global` is cool

File: template/src/containers/AuthContainer.tsx (hunk 8 above)
Comment (Harleymckee) [reply template/src/containers/AuthContainer.tsx]: Yeah that's my biggest TODO on here, I am going to try to nail that down this eve.

I got caught up in trying to implement a full oauth2 auth loop, but we should just start with encrypted storage.

File: template/src/navigation/SettingsStack.tsx (hunk 5 above)
Comment (Harleymckee) [reply template/src/navigation/SettingsStack.tsx]: nice

**Comment by Harleymckee** [comment]
re components, thinking we should have those live here https://github.com/revelrylabs/Harmonium-RN/ thoughts?

**Comment by Dacello** [comment]
@Harleymckee that makes sense

**Comment by Harleymckee** [comment]
Another TODO is to get a base Fastlane template going (base on https://github.com/revelrylabs/sunpro_mobile/blob/develop/fastlane/Fastfile).

That's something I am proud of on that project. I think I can make it more reusable too, with some manual steps to complete in app stores clear in the docs

**Comment by Dacello** [comment]
Theres some useful scripts that @brianberlin added here https://github.com/revelrylabs/TSS/tree/develop/bin/mobile, specifically codepush that might be useful in the future

**Comment by Harleymckee** [comment]
I havnt had to use codepush at all on Sunpro @Dacello, my deploys work fine just fastlane + xcodebuild/gradle build (which both have a bundle JS task), what does it do?

**Comment by Harleymckee** [comment]
here is sunpro deploy in total https://github.com/revelrylabs/sunpro_mobile/blob/develop/.travis.yml

**Comment by Dacello** [comment]
@nschello and @brianberlin might be able to chime in there, but my understanding of the advantage of codepush is that it doesnt require you to submit new builds to apple/google play if all that changed was JS. You can push a new build to code push and it forces all users to update without having to go thru app review process

**Comment by Harleymckee** [comment]
ah I see, cool will consider that moving forward. even though review via fastlane is pretty quick these days, it may be a bottleneck on large projs.

**Comment by Harleymckee** [comment]
@Dacello would you be cool with me merging this as is and creating issues for all the outstanding tasks mentioned in this pr?

**Comment by Dacello** [comment]
@Harleymckee wfm

**Comment by Dacello** [comment]
@Harleymckee btw just noticing i dont have write access to this repo based on what my approval shows up as. Can we get me (and everyone else) that ?

**Comment by Dacello** [comment]
Access received. Thanks

**Pull request merged**